File: design-patterns-ou-padroes-de-projeto/creational-patterns/singleton/singleton.cs
using System;

public sealed class Logger
{
    private static readonly Logger _instance = new Logger();

    // Construtor privado impede instanciamento externo
    private Logger() {}

    public static Logger Instance => _instance;

    public void Log(string message)
    {
        Console.WriteLine($"LOG: {message}");
    }
}

// Uso do Singleton
class Program
{
    static void Main()
    {
        var logger1 = Logger.Instance;
        var logger2 = Logger.Instance;

        logger1.Log("Isso é um Singleton!");

        Console.WriteLine(Object.ReferenceEquals(logger1, logger2));  // true
    }
}

File: design-patterns-ou-padroes-de-projeto/creational-patterns/singleton/singleton.py
class Logger:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Logger, cls).__new__(cls)
        return cls._instance

    def log(self, message):
        print(f"LOG: {message}")

# Uso do Singleton
logger1 = Logger()
logger2 = Logger()

logger1.log("Isso é um Singleton!")

print(logger1 is logger2)  # True (mesma instância)
